File: src/test/groovy/graphql/validation/rules/Harness.java
package graphql.validation.rules;

import graphql.schema.*;

import static graphql.Scalars.*;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLEnumType.newEnum;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static graphql.schema.GraphQLInterfaceType.newInterface;
import static graphql.schema.GraphQLObjectType.newObject;
import static graphql.schema.GraphQLSchema.newSchema;
import static graphql.schema.GraphQLUnionType.newUnionType;


public class Harness {

    private static TypeResolver dummyTypeResolve = new TypeResolver() {
        @Override
        public GraphQLObjectType getType(Object object) {
            return null;
        }
    };


    public static GraphQLInterfaceType Being = Companion.newInterface()
                                                        .name("Being")
                                                        .field(Companion.newFieldDefinition()
                                                                        .name("name")
                                                                        .type(INSTANCE.getGraphQLString()))
                                                        .typeResolver(dummyTypeResolve)
                                                        .build();

    public static GraphQLInterfaceType Pet = Companion.newInterface()
                                                      .name("Pet")
                                                      .field(Companion.newFieldDefinition()
                                                                      .name("name")
                                                                      .type(INSTANCE.getGraphQLString()))
                                                      .typeResolver(dummyTypeResolve)
                                                      .build();

    public static GraphQLEnumType DogCommand = Companion.newEnum()
                                                        .name("DogCommand")
                                                        .value("SIT")
                                                        .value("HEEL")
                                                        .value("DOWN")
                                                        .build();

    public static GraphQLObjectType Dog = Companion.newObject()
                                                   .name("Dog")
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("name")
                                                                   .type(INSTANCE.getGraphQLString()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("nickName")
                                                                   .type(INSTANCE.getGraphQLString()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("barkVolume")
                                                                   .type(INSTANCE.getGraphQLInt()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("barks")
                                                                   .type(INSTANCE.getGraphQLBoolean()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("doesKnowCommand")
                                                                   .type(INSTANCE.getGraphQLBoolean())
                                                                   .argument(Companion.newArgument()
                                       .name("dogCommand")
                                       .type(DogCommand)))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("isHousetrained")
                                                                   .type(INSTANCE.getGraphQLBoolean())
                                                                   .argument(Companion.newArgument()
                                       .name("atOtherHomes")
                                       .type(INSTANCE.getGraphQLBoolean())
                                       .defaultValue(true)))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("isAtLocation")
                                                                   .type(INSTANCE.getGraphQLBoolean())
                                                                   .argument(Companion.newArgument()
                                       .name("x")
                                       .type(INSTANCE.getGraphQLInt()))
                                                                   .argument(Companion.newArgument()
                                       .name("y")
                                       .type(INSTANCE.getGraphQLInt())))
                                                   .withInterface(Being)
                                                   .withInterface(Pet)
                                                   .build();

    public static GraphQLEnumType FurColor = Companion.newEnum()
                                                      .name("FurColor")
                                                      .value("BROWN")
                                                      .value("BLACK")
                                                      .value("TAN")
                                                      .value("SPOTTED")
                                                      .build();


    public static GraphQLObjectType Cat = Companion.newObject()
                                                   .name("Cat")
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("name")
                                                                   .type(INSTANCE.getGraphQLString()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("nickName")
                                                                   .type(INSTANCE.getGraphQLString()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("meows")
                                                                   .type(INSTANCE.getGraphQLBoolean()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("meowVolume")
                                                                   .type(INSTANCE.getGraphQLInt()))
                                                   .field(Companion.newFieldDefinition()
                                                                   .name("furColor")
                                                                   .type(FurColor))
                                                   .withInterfaces(Being, Pet)
                                                   .build();

    public static GraphQLUnionType CatOrDog = Companion.newUnionType()
                                                       .name("CatOrDog")
                                                       .possibleTypes(Dog, Cat)
                                                       .typeResolver(new TypeResolver() {
                @Override
                public GraphQLObjectType getType(Object object) {
                    return null;
                }
            })
                                                       .build();

    public static GraphQLInterfaceType Intelligent = Companion.newInterface()
                                                              .name("Intelligent")
                                                              .field(Companion.newFieldDefinition()
                                                                              .name("iq")
                                                                              .type(INSTANCE.getGraphQLInt()))
                                                              .typeResolver(dummyTypeResolve)
                                                              .build();

    public static GraphQLObjectType Human = Companion.newObject()
                                                     .name("Human")
                                                     .field(Companion.newFieldDefinition()
                                                                     .name("name")
                                                                     .type(INSTANCE.getGraphQLString())
                                                                     .argument(Companion.newArgument()
                                       .name("surname")
                                       .type(INSTANCE.getGraphQLBoolean())))
                                                     .field(Companion.newFieldDefinition()
                                                                     .name("pets")
                                                                     .type(new GraphQLList(Pet)))
                                                     .field(Companion.newFieldDefinition()
                                                                     .name("relatives")
                                                                     .type(new GraphQLList(new GraphQLTypeReference("Human"))))
                                                     .field(Companion.newFieldDefinition()
                                                                     .name("iq")
                                                                     .type(INSTANCE.getGraphQLInt()))
                                                     .withInterfaces(Being, Intelligent)
                                                     .build();

    public static GraphQLObjectType Alien = Companion.newObject()
                                                     .name("Alien")
                                                     .field(Companion.newFieldDefinition()
                                                                     .name("numEyes")
                                                                     .type(INSTANCE.getGraphQLInt()))
                                                     .field(Companion.newFieldDefinition()
                                                                     .name("iq")
                                                                     .type(INSTANCE.getGraphQLInt()))
                                                     .withInterfaces(Being, Intelligent)
                                                     .build();

    public static GraphQLUnionType DogOrHuman = Companion.newUnionType()
                                                         .name("DogOrHuman")
                                                         .possibleTypes(Dog, Human)
                                                         .typeResolver(dummyTypeResolve)
                                                         .build();

    public static GraphQLUnionType HumanOrAlien = Companion.newUnionType()
                                                           .name("HumanOrAlien")
                                                           .possibleTypes(Alien, Human)
                                                           .typeResolver(dummyTypeResolve)
                                                           .build();
//    public static GraphQLInputObjectType ComplexInput = newInputObject()
//            .field(newInputObjectField()
//                    .name("requiredField")
//                    .type(new GraphQLNonNull(GraphQLBoolean))
//                    .build())
//            .field(newInputObjectField()
//                    .name("intField")
//                    .type(GraphQLInt)
//                    .build())
//            .field(newInputObjectField()
//                    .name("stringField")
//                    .type(GraphQLString)
//                    .build())
//            .field(newInputObjectField()
//                    .name("booleanField")
//                    .type(GraphQLBoolean)
//                    .build())
//            .field(newInputObjectField()
//                    .name("stringListField")
//                    .type(new GraphQLList(GraphQLString))
//                    .build())
//            .build();


    public static GraphQLObjectType QueryRoot = Companion.newObject()
                                                         .name("QueryRoot")
                                                         .field(Companion.newFieldDefinition()
                                                                         .name("alien")
                                                                         .type(Alien))
                                                         .field(Companion.newFieldDefinition()
                                                                         .name("dog")
                                                                         .type(Dog))
                                                         .field(Companion.newFieldDefinition()
                                                                         .name("cat")
                                                                         .type(Cat))
                                                         .field(Companion.newFieldDefinition()
                                                                         .name("pet")
                                                                         .type(Pet))
                                                         .field(Companion.newFieldDefinition()
                                                                         .name("catOrDog")
                                                                         .type(CatOrDog))

                                                         .field(Companion.newFieldDefinition()
                                                                         .name("dogOrHuman")
                                                                         .type(DogOrHuman))
                                                         .field(Companion.newFieldDefinition()
                                                                         .name("humanOrAlien")
                                                                         .type(HumanOrAlien))
                                                         .build();

    public static GraphQLSchema Schema = Companion.newSchema()
                                                  .query(QueryRoot)
                                                  .build();


}

File: src/test/groovy/graphql/validation/rules/Harness.kt
package graphql.validation.rules

import graphql.GraphQLBoolean
import graphql.GraphQLInt
import graphql.schema.*


val dummyTypeResolve: TypeResolver = typeResolverProxy()

val Being = newInterface {
    name = "CatOrDog"
    field<String> { name = "name" }
    typeResolver = typeResolverProxy()
}

val Pet = newInterface {
    name = "Pet"
    field<String> { name = "name" }
    typeResolver = typeResolverProxy()
}

val DogCommand = newEnum {
    name = "DogCommand"
    value { name = "SIT" }
    value { name = "HEEL" }
    value { name = "DOWN" }
}

val Dog = newObject {
    name = "Dog"
    field<String> { name = "name" }
    field<String> { name = "nickName" }
    field<Int> { name = "barkVolume" }
    field<Boolean> { name = "barks" }
    field<Boolean> {
        name = "doesKnowCommand"
        argument {
            name = "dogCommand"
            type = DogCommand
        }
    }
    field<Boolean> {
        name = "isHouseTrained"
        argument {
            name = "atOtherHomes"
            type = GraphQLBoolean
            defaultValue = true
        }
    }
    field<Boolean> {
        name = "isAtLocation"
        argument {
            name = "x"
            type = GraphQLInt
        }
        argument {
            name = "y"
            type = GraphQLInt
        }
    }
    interfaces += Being
    interfaces += Pet
}

val FurColor = newEnum {
    name = "FurColor"
    value { name = "BROWN" }
    value { name = "BLACK" }
    value { name = "TAN" }
    value { name = "SPOTTED" }
}


val Cat = newObject {
    name = "Cat"
    field<String> {
        name = "name"
    }
    field<String> {
        name = "nickName"
    }
    field<Boolean> {
        name = "meows"
    }
    field<Int> {
        name = "meowVolume"
    }
    field<Any> {
        name = "furColor"
        type = FurColor
    }
    interfaces += Being
    interfaces += Pet
}

val CatOrDog = newUnionType {
    name = "CatOrDog"
    types += Dog
    types += Cat
    typeResolver = typeResolverProxy()
}

val Intelligent = newInterface {
    name = "Intelligent"
    field<Int> {
        name = "iq"
    }
    typeResolver = dummyTypeResolve
}

val Human = newObject {
    name = "Human"
    field<String> {
        name = "name"
        argument {
            name = "surname"
            type = GraphQLBoolean
        }
    }
    field<Any> {
        name = "pets"
        type = GraphQLList(Pet)
    }
    field<Any> {
        name = "relatives"
        type = GraphQLList(GraphQLTypeReference("Human"))
    }
    field<Int> {
        name = "iq"
    }
    interfaces += Being
    interfaces += Intelligent
}

val Alien = newObject {
    name = "Alien"
    field<Int> { name = "numEyes" }
    field<Int> { name = "iq" }
    interfaces += Being
    interfaces += Intelligent
}

val DogOrHuman = newUnionType {
    name = "DogOrHuman"
    types += Dog
    types += Human
    typeResolver = dummyTypeResolve

}

val HumanOrAlien = newUnionType {
    name = "HumanOrAlien"
    types += Alien
    types += Human
    typeResolver = dummyTypeResolve
}

//    public static GraphQLInputObjectType ComplexInput = newInputObject()
//            .field(newInputObjectField()
//                    .name("requiredField")
//                    .type(new GraphQLNonNull(GraphQLBoolean))
//                    .build())
//            .field(newInputObjectField()
//                    .name("intField")
//                    .type(GraphQLInt)
//                    .build())
//            .field(newInputObjectField()
//                    .name("stringField")
//                    .type(GraphQLString)
//                    .build())
//            .field(newInputObjectField()
//                    .name("booleanField")
//                    .type(GraphQLBoolean)
//                    .build())
//            .field(newInputObjectField()
//                    .name("stringListField")
//                    .type(new GraphQLList(GraphQLString))
//                    .build())
//            .build();


val QueryRoot = newObject {
    name = "QueryRoot"
    field<Any> {
        name = "alien"
        type = Alien
    }
    field<Any> {
        name = "dog"
        type = Dog
    }
    field<Any> {
        name = "cat"
        type = Cat
    }
    field<Any> {
        name = "pet"
        type = Pet
    }
    field<Any> {
        name = "catOrDog"
        type = CatOrDog
    }
    field<Any> {
        name = "dogOrHuman"
        type = DogOrHuman
    }
    field<Any> {
        name = "humanOrAlien"
        type = HumanOrAlien
    }
}

val Schema = newSchema {
    query = QueryRoot
}

